# Hypothesis Test of the Means:




---
>* [t Tests for One Mean: Introduction](https://www.youtube.com/watch?v=T9nI6vhTU1Y&t=576s)
>* [Hypothesis tests on one mean: t test or z test?](https://www.youtube.com/watch?v=vw2IPZ2aD-c)


---



some points to keep in mind about p-value and hypothesis testing in general: 

• **The interpretation of the p-value :** 

it represents the prob of having a sample mean greater then the null value of H0 or, from  n random samples, given that the H0 is true  ( a conditionned probability ) 
In other words 
        P-value = P(extreme events in favor to alternative hypothesis  | H0 is true ) 



• High P-values: Your sample results are consistent with a true null hypothesis. <br>
• Low P-values: Your sample results are not consistent with a null hypothesis. <br>

• The significance level, also denoted as alpha or α, is the probability of rejecting the null  hypothesis when it is true


• Type 1 error : reject the H0 when the latter is true <br>
• Type 2 error : accept the H0 when the latter is false , in other words , we keep the H0 when we shouldn't <br>


• Type 1 error expressed with propability : 
  P-value = P(typ1 error |H0 is true ) = alpha 
  Increasing alpha -> increase error rate of making a type I error rate!



• **Goal of Increasing alpha (increase type 1 error) :** when we don't  want to fail to reject the H0 when the latter is false <br>
• **Goal of reducing alpha :** when we want to reject the H0 ONLY if we have STRONG evidence  favoring the Ha.



## One sample t-test:

In [ ]:
import numpy as np
import scipy.stats as st

**Two-side test**

Question 1 : the goal of this exercice is to simulate one sample x with normal distribution

- step 1 : create n=10 random values using `np.random.randn` , note that this instruction returns z scores of standard normal
- step 2 : do the inverse transform to get x from these z scores using mu=3 and sigma=2 ( hint : use the forumla of z score ) and store the result in variable x
\begin{equation}
 z= \frac{x-\mu}{\sigma}
\end{equation}


- step 3 : we wish to test if the null hypothesis with a **two-side alternative hypothesis** , run the following and interpret the result , knowing that mu_0 is the mean of the null hypothesis 

``` 
mu_0 = 0
st.ttest_1samp(x, mu_0)

```

- step 4 : forumate the associated null hypothesis $H_{0}$ and alternative hypothesis $H_{a}$


In [ ]:
#don't delete this line 
np.random.seed(1234)
z=np.random.randn(10)
mu=3
sigma=2
x=z*sigma+mu
#H0:µ=0
#H1:µ!=0
mu_0 = 0
st.ttest_1samp(x, mu_0)


Ttest_1sampResult(statistic=3.8682023385794677, pvalue=0.0037991794351627216)

>The statistic is the t value we get, and the p value is the corresponding probability of having a value as or more extreme under the null hypothesis

Question 2 : now that you know how to use `st.ttest_1samp` , use the forumlas that you learnt during the course to get the same results , here's the step to follow : 

- step 1 : calculate the `mean` and the `standard error` $SE$ 
- step 2 : calculate the t-score using the following forumla :


\begin{equation}
  t = \frac{x-\bar{x} }{ \frac{S}{\sqrt{n} } }
\end{equation}

- step 3 : calculate the probability of `p-value` using `st.t.cdf` , remember that p-value = $2*P(X>tscore)$ 
  



In [ ]:
mu_0=0
n=10
mean=np.mean(x)

SE=(np.std(x,ddof=1))/(np.sqrt(10))

t=(mean-mu_0)/SE

p_value=2*(1-st.t.cdf(t,n-1))
t,p_value

(3.8682023385794677, 0.003799179435162836)

Question 3: same as the above but with `mu_0 = 1` , `mu_0 = 2` 


In [ ]:
mu_0=1
n=10
mean=np.mean(x)

SE=(np.std(x,ddof=1))/(np.sqrt(10))

t=(mean-mu_0)/SE

p_value=2*(1-st.t.cdf(t,n-1))
t,p_value

(2.4422068880199026, 0.03722962101247096)

In [ ]:
mu_0=2
n=10
mean=np.mean(x)

SE=(np.std(x,ddof=1))/(np.sqrt(10))

t=(mean-mu_0)/SE

p_value=2*(1-st.t.cdf(t,n-1))
t,p_value

(1.0162114374603375, 0.3360674694425527)

**right-tail test**

Question 4 : everything remain the same , except that the p-value is calculated differently , since : 

 p-value = $P(X>tscore)$ 


calculate the p-value with mu_0= 0 , mu_0= 1 , mu_0= 2 

In [ ]:
mu_0=0
n=10
mean=np.mean(x)

SE=(np.std(x,ddof=1))/(np.sqrt(10))

t=(mean-mu_0)/SE

p_value=1-st.t.cdf(t,n-1)
t,p_value

(3.8682023385794677, 0.001899589717581418)

In [ ]:
mu_0=1
n=10
mean=np.mean(x)

SE=(np.std(x,ddof=1))/(np.sqrt(10))

t=(mean-mu_0)/SE

p_value=1-st.t.cdf(t,n-1)
t,p_value

(2.4422068880199026, 0.01861481050623548)

In [ ]:
mu_0=2
n=10
mean=np.mean(x)

SE=(np.std(x,ddof=1))/(np.sqrt(10))

t=(mean-mu_0)/SE

p_value=1-st.t.cdf(t,n-1)
t,p_value

(1.0162114374603375, 0.16803373472127636)

**left-tail test**

Question 5 : everything remain the same , except that the p-value is calculated differently , since : 

 p-value = $P(X<tscore)$ 


calculate the p-value with mu_0= 0 , mu_0= 1 , mu_0= 2 

In [ ]:
mu_0=0
n=10
mean=np.mean(x)

SE=(np.std(x,ddof=1))/(np.sqrt(10))

t=(mean-mu_0)/SE

p_value=st.t.cdf(t,n-1)
t,p_value

(3.8682023385794677, 0.9981004102824186)

we have no evidence to reject the null hypothesis stating that µ=0

In [ ]:
mu_0=1
n=10
mean=np.mean(x)

SE=(np.std(x,ddof=1))/(np.sqrt(10))

t=(mean-mu_0)/SE
p_value=st.t.cdf(t,n-1)
t,p_value

(2.4422068880199026, 0.9813851894937645)

In [ ]:
mu_0=0
n=10
mean=np.mean(x)

SE=(np.std(x,ddof=1))/(np.sqrt(10))

t=(mean-mu_0)/SE

p_value=st.t.cdf(t,n-1)
t,p_value

(3.8682023385794677, 0.9981004102824186)

µ is surely not below 0

## independent two sample t-test

---
> The null hypothesis is that µ1=µ2
---

Question 6 : just like you created your samples in the first question , follow the same steps in order to create x1 and x2 that have the following propreties :

x1 (sample 1 ) : n1 = 10 , mu1 = 3 , sigma1 = 2

x2 (sample 2) :  n2 = 20 , mu2 = 5 , sigma2 = 3

In [ ]:
#make sure to name your variables x1 and x2
n1 = 10 
mu1 = 3 
sigma1 = 2
n2 = 20 
mu2 = 5
sigma2 = 3
x1=(np.random.randn(n1))*sigma1+mu1
x2=(np.random.randn(n2))*sigma2+mu2
x1,x2

(array([ 5.30007145,  4.98389204,  4.90664826, -1.04250964,  2.33184527,
         3.00423673,  3.81090682,  3.57818388,  5.64231638, -0.09381111]),
 array([ 4.39206103,  3.03209197,  5.58026413,  6.66031673,  8.95445466,
         3.59208415,  7.02666226, -0.45108168,  4.45067438,  8.17690756,
         3.80647932,  6.01231296,  8.14273572,  8.13781477,  7.59115188,
         4.63372528,  5.37413886,  4.03161558,  7.52502414, 12.17288155]))

Question 7 : run the following and interpret the results : 

In [ ]:
st.ttest_ind(x1,x2,equal_var=False)                                       

Ttest_indResult(statistic=-2.870701133767877, pvalue=0.009090146778647546)

The result show that is not so probable to get such a difference between the two sample means if really they had the same population mean (if they came from the same population). Indeed, p_value=2%

Question 8 : calculate x1_mean , x2_mean  , s1 (unbiased standard deviation) , s2 of the two samples x1 and x2

In [ ]:
x1_mean=np.mean(x1)
s1=np.std(x1,ddof=1)
print(x1_mean,s1)
x2_mean=np.mean(x2)
s2=np.std(x2,ddof=1)
print(x2_mean,s2)

3.2421780090611434 2.2756846201082483
5.942115761217228 2.708115937753849


Question 9 : run the function using the descriptive statistics optained in the previous question using  `st.ttest_ind_from_stats ` (i invite you to read the documentation first  ) , also make sure to set `equal_var=False`    

In [ ]:
t2, p2 = st.ttest_ind_from_stats(x1_mean, s1, n1,
                              x2_mean, s2, n2,
                              equal_var=False)
t2,p2

(-2.870701133767877, 0.009090146778647546)

## Paired two sample t-test

---
>* [Comparative Statistics in Python using SciPy](https://benalexkeen.com/comparative-statistics-in-python-using-scipy/)
>* [T-Tests: A Matched Pair Made in Heaven: Crash Course Statistics #27](https://www.youtube.com/watch?v=AGh66ZPpOSQ&list=PL8dPuuaLjXtNM_Y-bUAhblSAdWRnmBUcr&index=28)
---

Question 10 : Simulate two samples of size **n=10** with Normal distribution. to do so : 

- step 1 : create x1 using the same method as described in question1 with mu=3 , sigma=2
- step 2 : create a "noisy" distribution derived from x1 using the following formula : 
                    x2 = x1 + np.random.randn(n) + 1
                    
(Notice that the x2 is created by adding some 'noise' to the x1)

So,'one to one' relation is established.

In [ ]:
x1=(np.random.randn(10))*2+3
x2 = x1 + np.random.randn(10) 


Question 11 : 
run the following line and interpret the results 


In [ ]:
# Two tail test.
st.ttest_rel(x1,x2)

Ttest_relResult(statistic=0.12497116409784303, pvalue=0.9032932801513642)

the two samples are originally the same, with juste some noise, thus, the sample means are not too different, this explains that we get a p_value of 37% in favour of the null hypothesis (µ1=µ2)

Question 12 : Compare with the indepdent two sample t-test on the same dataset. 


In [ ]:
st.ttest_ind(x1,x2,equal_var=False)  

Ttest_indResult(statistic=0.045656103959977164, pvalue=0.9641170935300101)

We notice that the p_value in the case of paired t-test is smaller than the p_value for the independant t-test
Paired T-tests are used when we are taking repeated samples and want to take into account the fact that the two distributions we are testing are paired.



in the paired t-test, instead of calculating the difference of the two sample means, we calculate the mean of the diffrences between the paired sample

## ANOVA

---
> Used to compare 3 samples 
>* [Analysis of Variance (ANOVA)](https://www.youtube.com/watch?v=ITf4vHhyGpc&t=26s)
---

There are three groups of students. The data corresponds to the test scores. Determine whether there is any significant difference in the group test scores.

In [ ]:
# Data.
x1 = np.array([82,93,61,74,69,70,53])
x2 = np.array([71,62,85,94,78,66,71])
x3 = np.array([64,73,87,91,56,78,87])

In [ ]:
# Using the f_onway() function from the SciPy library.
res=st.f_oneway(x1,x2,x3)
print("Test statistic  : %f " % res[0] )
print("p-value         : %f " % res[1] )

Test statistic  : 0.283726 
p-value         : 0.756278 


There is not a significant difference in the group test scores 

# Summarizing categorical variables:

In [ ]:
import pandas as pd
df = pd.read_csv('data_studentlist.csv', header='infer',encoding = 'latin1')

## Frequency table (One way table):

In [ ]:
# Categorical variables.
bloodType = df['bloodtype']
gender = df['gender']

get the frequency table of bloodtype

In [ ]:
bloodType.value_counts()

B     5
O     5
A     4
AB    3
Name: bloodtype, dtype: int64

get the frequency table of gender

In [ ]:
gender.value_counts()

M    10
F     7
Name: gender, dtype: int64

get the normalized frequency table of bloodtype

In [ ]:
bloodType.value_counts(normalize=True)

B     0.294118
O     0.294118
A     0.235294
AB    0.176471
Name: bloodtype, dtype: float64

same as the above for gender

In [ ]:
gender.value_counts(normalize=True)

M    0.588235
F    0.411765
Name: gender, dtype: float64

##  Contingency table (Two way table):

get the Contingency table using `pd.crosstab`

In [ ]:
pd.crosstab(bloodType,gender)

gender     F  M
bloodtype      
A          2  2
AB         1  2
B          1  4
O          3  2

in order to get Contingency table with marginalization , set the parameter `margins=True`


In [ ]:
pd.crosstab(bloodType,gender,margins=True)

gender     F   M  All
bloodtype            
A          2   2    4
AB         1   2    3
B          1   4    5
O          3   2    5
All        7  10   17

# Hypothesis Test of proportions:
 

## Chi-squared Tests:

---
>* [Chi-Square Tests: Crash Course Statistics #29](https://www.youtube.com/watch?v=7_cs1YlZoug&list=PL8dPuuaLjXtNM_Y-bUAhblSAdWRnmBUcr&index=30&t=548s)
---

###  For one way table 

Suppose we have a frequency table of *eye color* of 592 adults. <br>
Let us carry out the Chi-squared tests.

In [ ]:
# Data.
data = np.array([220, 215, 93,64])

- step 1  : create a series with data , and set the table `['Brown','Blue','Hazel','Green']` as index   

In [ ]:
colors=['Brown','Blue','Hazel','Green']
eye=pd.Series(data,index=colors)
eye

Brown    220
Blue     215
Hazel     93
Green     64
dtype: int64

Question 13 : create a Table of proportions. to do so , divide your series by the sum of the frequencies

In [ ]:
eye_prop=eye/data.sum()

Question 14 : Compare with the default model of equal frequencies. ( in other words , calculate the chisquare of your series using `st.chisquare`

In [ ]:
st.chisquare(eye)

Power_divergenceResult(statistic=133.47297297297297, pvalue=9.650879774148268e-29)

that is a really really really small p_value, meaning that if the frequencies of eye color are really equal in the population, it is very unlikely that we observe the frequencies we obsered in our sample

Question 15 : the following is a model provided by the user that tries to approximate the frequencies 

In [ ]:
my_model = len(eye)*[np.mean(eye)]

run the following and describe the result :

In [ ]:
st.chisquare(f_obs=eye,f_exp = my_model)

Power_divergenceResult(statistic=133.47297297297297, pvalue=9.650879774148268e-29)

we get the same result as above because by default, in the chisquare function the categories are assumed to be equally likely.

let's test with a more realistic model : 

In [ ]:
my_model = eye.sum()*np.array([0.40,0.33,0.15,0.12])

run the following and describe the result. what do you conlcude about the two proposed models ? 

In [ ]:
st.chisquare(f_obs=eye,f_exp = my_model )

Power_divergenceResult(statistic=4.06265356265356, pvalue=0.25477767063672113)

Indeed, this model is more realistic because the p value is equal to 25%

###  For two way table:

Suppose we have a contingency table of *eye color* vs *hair color* of 592 adults. <br>
Let us carry out the Chi-squared test for independence.

In [ ]:
# Data.
# Row label    = hair color.
# Column label = eye color. 
data_eye = {'Brown': [68,119,26,7], 'Blue':[20,84,17,94], 'Hazel':[15,54,14,10], 'Green':[5, 29, 14,16]}
data = pd.DataFrame(data_eye,index=['Black','Brown','Red','Blonde'])
data

Brown  Blue  Hazel  Green
Black      68    20     15      5
Brown     119    84     54     29
Red        26    17     14     14
Blonde      7    94     10     16

run the following to get Chi-square test of independence of variables in a contingency table. what do you conclude about the relationship between these two categorical variables ? 

In [ ]:
chi2, p, dof, expected_frequencies = st.chi2_contingency(data)
print("expected frequencies based on the marginal sums of the table:\n",expected_frequencies)
print('p_value',p)

expected frequencies based on the marginal sums of the table:
 [[ 40.13513514  39.22297297  16.96621622  11.67567568]
 [106.28378378 103.86824324  44.92905405  30.91891892]
 [ 26.38513514  25.78547297  11.15371622   7.67567568]
 [ 47.19594595  46.12331081  19.95101351  13.72972973]]
p_value 2.325286787098839e-25


In [ ]:
data['Tot']=data['Brown']+data['Blue']+data['Hazel']+data['Green']
data.loc[4]=data.loc['Black']+data.loc['Brown']+data.loc['Red']+data.loc['Blonde']

In [ ]:
data

Brown  Blue  Hazel  Green  Tot
Black      68    20     15      5  108
Brown     119    84     54     29  286
Red        26    17     14     14   71
Blonde      7    94     10     16  127
4         220   215     93     64  592

the expected frequencies are calculated this way : 

for example, for the cell where hair=black and eye color=brown, we multiply the relative frequency of eye=brown by the total number of hair=black: (220/592)*108 and we get 40.135...


the p value is really small, meaning that the two variables (hair and eye color) must be dependent

Suppose we have a contingency table of the room class vs survived from RMS Titanic.
Let us carry out the Chi-squared test for independence.

In [ ]:
# Data.
# Row label    = room class.
# Column label = survived. 
data_survived = {'Yes':[122, 167, 528, 673], 'No':[203,118,178,212]}
data = pd.DataFrame(data_survived,index=['1st','2nd','3rd','Crew'])
data

Yes   No
1st   122  203
2nd   167  118
3rd   528  178
Crew  673  212

run the following to get Chi-square test of independence of variables in a contingency table. what do you conclude about the relationship between these two categorical variables ? 

In [ ]:
st.chi2_contingency(data)

(138.28984162600824,
 1.5567852582894594e-21,
 16,
 array([[ 40.13513514,  39.22297297,  16.96621622,  11.67567568,
         108.        ],
        [106.28378378, 103.86824324,  44.92905405,  30.91891892,
         286.        ],
        [ 26.38513514,  25.78547297,  11.15371622,   7.67567568,
          71.        ],
        [ 47.19594595,  46.12331081,  19.95101351,  13.72972973,
         127.        ],
        [220.        , 215.        ,  93.        ,  64.        ,
         592.        ]]))

from the pvalue, we can conclude that the two variables are dependent